src/CachePool.php

```diff
@@ -0,0 +1,388 @@
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function clear()
+    {
+        try {
+            $keys = $this->getAllKeys();
+            $this->deleteMultiple($keys);
+        } catch (Exception $e) {
+            throw new CacheException(sprintf('Could not clear cache'), 0, $e);
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getMultiple($keys, $default = null)
+    {
+        if (!is_iterable($keys)) {
+            throw new InvalidArgumentException(sprintf('List of keys is not a list'));
+        }
+
+        $entries = [];
+        foreach ($keys as $key) {
+            $value = $this->get($key, $default);
+            $entries[$key] = $value;
+        }
+
+        return $entries;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function setMultiple($values, $ttl = null)
+    {
+        if (!is_iterable($values)) {
+            throw new InvalidArgumentException(sprintf('List of keys is not a list'));
+        }
+
+        foreach ($values as $key => $value) {
+            $this->set($key, $value, $ttl);
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function deleteMultiple($keys)
+    {
+        if (!is_iterable($keys)) {
+            throw new InvalidArgumentException(sprintf('List of keys is not a list'));
+        }
+
+        foreach ($keys as $key) {
+            $this->delete($key);
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function has($key)
+    {
+        $default = $this->defaultValue;
+        $prefix = $this->getOptionNamePrefix();
+        $value = $this->getOption("{$prefix}{$key}", $default);
+
+        return $value !== $default;
+    }
+
+    /**
+     * Retrieves a transient value, by key.
+     *
+     * @param string $key The transient key.
+     *
+     * @return string|bool The transient value.
+     */
+    protected function getTransient(string $key)
+    {
+        $value = get_transient($key);
+
+        return $value;
+    }
+
+    /**
+     * Retrieves an option value by name.
+     *
+     * @param string $name    The option name.
+     * @param null   $default The value to return if option not found.
+     *
+     * @return string The option value.
+     */
+    protected function getOption(string $name, $default = null): string
+    {
+        return (string) get_option($name, $default);
+    }
+
+    /**
+     * Validates a cache key.
+     *
+     * @param string $key The key to validate.
+     *
+     * @throws InvalidArgumentException If key is invalid.
+     */
+    protected function validateKey(string $key)
+    {
+        $reservedSymbols = str_split(static::RESERVED_KEY_SYMBOLS, 1);
+
+        foreach ($reservedSymbols as $symbol) {
+            if (strpos($key, $symbol) !== false) {
+                throw new InvalidArgumentException(sprintf('Cache key "%1$s" is invalid', $key));
+            }
+        }
```

Comment:
A `preg_match()` would be much more efficient here:

```php
if (preg_match('/[' . static::RESERVED_KEY_SYMBOLS . ']/', $key) {
	throw new InvalidArgumentException(...);
}
```

Reply:
I don't actually know about efficiency comparison in this case. I was under the impression that regex is generally slow.

Reply:
Regex may be slow, but I think it would be better than a worst case of `O(n.m)` (where `n` is the number of reserved symbols and `m` is the length of the key).

Reply:
Dunno. I think probably in this case, since it's letter by letter, regex is doing the same thing as `str_split()`, but with more stuff around it.